native-addons-are-close-to-being-on-par-with-js-modules.md

```diff
@@ -0,0 +1,113 @@
+
+0. Node.js calculates the absolute path to the native addon.
+0. Node.js calls the method [`process.dlopen()`][], passing in the absolute path previously calculated.
+0. On the native side, Node.js calls [`dlopen(3)`][] or [`uv_dlopen()`][] on Windows.
+0. The addon is loaded and it executes a so-called DSO constructor function as part of the loading process. The function passes a pointer to a structure of type `node::nm_module` to Node.js.
+0. The `node::node_module::nm_register_func` or `node::node_module::nm_context_register_func` member of the structure holds a pointer to a function which is responsible for populating the object which will become the result of having loaded the module (`module.exports`).
+
+The problem with DSO constructor functions is that they only run the first time an addon is loaded into memory. On subsequent occasions, `dlopen(3)` or `uv_dlopen()` will short-circuit, returning a reference to an already loaded addon, and the function will not run. Thus, subsequent attempts to load the addon fail.
+
+The solution to this problem was introduced in [3828fc62][]. In addition to relying on a DSO constructor, as a fallback Node.js now looks for a well-known symbol exported by the module: `node_register_module_v<number>`, where `<number>` is an integer representing the version of the addon. It only does so if the DSO constructor fails to run. The well-known symbols is the address of the module initialization function that is also stored in `node::node_module::nm_context_register_func`.
+
+Although originally this change was made with the intention of allowing well-known symbols for multiple versions of Node.js to co-habitate within the same binary, it also enables multiple loading. Nevertheless, the ability to load a native addon multiple times means that the addon must adhere to a [context-aware addon][] structure.
+
+### 5. Thread Safety
+Since JavaScript modules are self-contained, even given their own scope at module load time by Node.js, they raise no thread safety concerns. In contrast, native addons must avoid global state in order to be considered thread-safe. This requires an explicit design decision on the part of the maintainer. Fortunately, per-addon-instance data can be heap-allocated at module initialization time, and passed to each native binding made available from the addon as seen in the example of the [context-aware addon][]. This applies to both N-API addons and V8 addons.
+
+
+## Conclusion
+Although the need for rendering a native addon as multiple binaries, one for each platform and architecture it supports, cannot be eliminated, with recent Node.js improvements, we can consider native addons as having taken a significant step towards being "on-par" with JavaScript modules:
+
+|  |  | JavaScript modules | Native addons then | Native addons now |
```

Comment:
Maybe change the header `Native addons now` to `Native addons now with N-API`?

Reply:
@jschlight ... with N-API, multiple, context-aware module loading, and proper module unloading ...

I think I'll change the title to "New Node.js Features Bring Native Add-ons Close To Being On-Par With JavaScript Modules"

Excellent idea to emphasize right off the bat that the article will be talking about new features!